### PR TITLE
state: Generalise error handling when splitting logs collection

### PR DESCRIPTION
**Pull request by mjs**
## Description of change

The error checks in SplitLogCollections didn't work on MongoDB 2.4 as
older MongoDB versions only return an error string without the error
code.

mgo's IsDup helper is now used for insert error handling and "ns not
found" is now checked by code or string.

Fixes https://bugs.launchpad.net/juju/+bug/1701999

## QA steps

Related unit tests were failing on MongoDB 2.4 before and pass now.

## Documentation changes

N.A. 

## Bug reference

https://bugs.launchpad.net/juju/+bug/1701999

### CODE REVIEW

**Comment by mjs** [comment]
$$merge$$

File: state/upgrades.go [hunk 1]
```diff
@@ -881,18 +877,30 @@ func SplitLogCollections(st *State) error {
...
 		}
 		return errors.Annotate(err, "failed to drop old logs collection")
 	}
 	return nil
 }
 
+func isMgoNamespaceNotFound(err error) bool {
+	// Check for &mgo.QueryError{Code:26, Message:"ns not found"}
+	if qerr, ok := err.(*mgo.QueryError); ok {
+		if qerr.Code == 26 {
```
Comment (jameinel) [inline state/upgrades.go]: Off-hand, shouldn't we have a a name for the error code instead of just a number mgo.NamespaceNotFoundCode or somesuch?

Comment (mjs) [inline state/upgrades.go]: No such constant exists in the mgo package. Is that what you mean?

**Comment by jameinel** [comment]
I mean that something like '26' should definitely be pulled into a
constant. Having it in 'mgo' would be ideal, but if not there, we should
have our own.
Its somewhat ok that we're at least using a helper
(isMgoNamespaceNotFound), but it still feels very odd to not have the
collection of error codes somewhere.

**Pull request merged**